improve remote image rendering

woodsaj opened the pull request:
- determine "domain" during Init() so we are not re-parsing settings
  on every request
- if using http-mode via a rednererUrl, then use the AppUrl for the page that the renderer loads.  When in http-mode the renderer is likely running on another server so trying to use the localhost or even the specific IP:PORT grafana is listening on wont work.
- apply the request timeout via a context rather than directly on the http client.
- use a global http client so we can take advantage of connection re-use
- log and handle errors better.

Commit:
improve remote image rendering

- determine "domain" during Init() so we are not re-parsing settings
  on every request
- if using http-mode via a rednererUrl, then use the AppUrl for the
  page that the renderer loads.  When in http-mode the renderer is likely
  running on another server so trying to use the localhost or even the
  specific IP:PORT grafana is listening on wont work.
- apply the request timeout via a context rather then directly on the http client.
- use a global http client so we can take advantage of connection re-use
- log and handle errors better.

Commit:
improve remote image rendering

- determine "domain" during Init() so we are not re-parsing settings
  on every request
- if using http-mode via a rednererUrl, then use the AppUrl for the
  page that the renderer loads.  When in http-mode the renderer is likely
  running on another server so trying to use the localhost or even the
  specific IP:PORT grafana is listening on wont work.
- apply the request timeout via a context rather then directly on the http client.
- use a global http client so we can take advantage of connection re-use
- log and handle errors better.

Commit:
ensure imagesDir exists

Commit:
ensure imagesDir exists

woodsaj commented:
i have also created https://github.com/grafana/grafana-image-renderer/pull/3  for testing this.

To test, just build the docker image from this branch
```
docker build -t grafana:dev .
```
then use the docker-compose in the grafana-image-renderer repo to bring up both grafana and the image-renderer HTTP service.

torkelo commented on pkg/services/rendering/rendering.go:
```diff
@@ -27,12 +29,31 @@ type RenderingService struct {
+	if rs.Cfg.RendererUrl != "" {
+		// AppUrl has already been passed, it wont generate an error.
+		u, _ := url.Parse(setting.AppUrl)
+		rs.domain = u.Hostname()
```
Think this will stop Grafana systems behind a reverse proxy to work. as the AppUrl is the public (reverse proxy url) and grafana does not have the credentials to go through that (for alerting and other scenarios) so need to access directly ad bypass auth proxy.

woodsaj replied:
That is true.  The only way to accommodate all use cases then is to add a separate `[rendering]` setting for the URL the renderer should load, ie `callback_url`.  If the `callback_url` is set to `""` (the default) we just use AppUrl.

Commit:
allow users to define callback_url for remote rendering

- allow users to define the url that a remote rendering service
  should use for connecting back to the grafana instance.
  By default the "root_url" is used.

Commit:
allow users to define callback_url for remote rendering

- allow users to define the url that a remote rendering service
  should use for connecting back to the grafana instance.
  By default the "root_url" is used.

torkelo committed:
rendering: fixed issue with renderKey where userId and orgId was in mixed up, added test for RenderCallbackUrl reading logic

torkelo committed:
Merge branch 'remoteRendering' of github.com:grafana/grafana into remoteRendering

torkelo commented:
had time to test this today, fixed issue with orgId and userId being mixed up in renderKey creation, not sure if that was bug in the master code or not (the plugin or http mode has not been used I think). 

Anyway tested grpc plugin mode & http mode, both with and without a reverse proxy infront. and tested alerting image generation as well. So looking good! merging when all is green. Oh, added unit test for the callback url setting parsing and slash handling.
